engine/platform/win32/con_win.c

```diff
@@ -25,197 +25,428 @@ WIN32 CONSOLE
+FOREGROUND_RED | FOREGROUND_GREEN,						// yellow
+FOREGROUND_BLUE | FOREGROUND_INTENSITY,					// blue
+FOREGROUND_GREEN | FOREGROUND_BLUE,						// cyan
+FOREGROUND_RED | FOREGROUND_BLUE,						// magenta
+FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE,	// default color (white)
+};
+
+static BOOL WINAPI Wcon_HandleConsole(DWORD CtrlType)
+{
+	return TRUE;
+}
+
+static void Wcon_PrintInternal( const char *msg, int length )
+{
+	char *pTemp;
+	DWORD cbWritten;
+	const char *pMsgString;
+	static char tmpBuf[2048];
+	static char szOutput[2048];
+
+	Q_strncpy( szOutput, msg, length ? (length + 1) : ( sizeof( szOutput ) - 1 ));
+	if( length )
+		szOutput[length + 1] = '\0';
+	else
+		szOutput[sizeof( szOutput ) - 1] = '\0';
+
+	pTemp = tmpBuf;
+	pMsgString = szOutput;
+	while( pMsgString && *pMsgString )
+	{
+		if( IsColorString( pMsgString ))
+		{
+			if (( pTemp - tmpBuf ) > 0 )
+			{
+				// dump accumulated text before change color
+				*pTemp = 0; // terminate string
+				WriteFile( s_wcd.hOutput, tmpBuf, Q_strlen(tmpBuf), &cbWritten, 0 );
+				pTemp = tmpBuf;
+			}
+
+			// set new color
+			SetConsoleTextAttribute( s_wcd.hOutput, g_color_table[ColorIndex(*(pMsgString + 1))] );
+			pMsgString += 2; // skip color info
+		}
+		else if(( pTemp - tmpBuf ) < sizeof( tmpBuf ) - 1 )
+		{
+			*pTemp++ = *pMsgString++;
+		}
+		else
+		{
+			// temp buffer is full, dump it now
+			*pTemp = 0; // terminate string
+			WriteFile( s_wcd.hOutput, tmpBuf, Q_strlen(tmpBuf), &cbWritten, 0 );
+			pTemp = tmpBuf;
+		}
+	}
+
+	// check for last portion
+	if (( pTemp - tmpBuf ) > 0 )
+	{
+		// dump accumulated text
+		*pTemp = 0; // terminate string
+		WriteFile( s_wcd.hOutput, tmpBuf, Q_strlen(tmpBuf), &cbWritten, 0 );
+		pTemp = tmpBuf;
+	}
+
+	// restore white color
+	SetConsoleTextAttribute( s_wcd.hOutput, g_color_table[7] );
+}
 
 void Wcon_ShowConsole( qboolean show )
 {
-	if( !s_wcd.hWnd || show == s_wcd.status )
+	if( !s_wcd.hWnd || show == s_wcd.consoleVisible )
 		return;
 
-	s_wcd.status = show;
+	s_wcd.consoleVisible = show;
 	if( show )
-	{
-		ShowWindow( s_wcd.hWnd, SW_SHOWNORMAL );
-		SendMessage( s_wcd.hwndBuffer, EM_LINESCROLL, 0, 0xffff );
-	}
-	else ShowWindow( s_wcd.hWnd, SW_HIDE );
+		ShowWindow( s_wcd.hWnd, SW_SHOW );
+	else
+		ShowWindow( s_wcd.hWnd, SW_HIDE );
 }
 
 void Wcon_DisableInput( void )
 {
-	if( host.type != HOST_DEDICATED ) return;
-	SendMessage( s_wcd.hwndButtonSubmit, WM_ENABLE, 0, 0 );
-	SendMessage( s_wcd.hwndInputLine, WM_ENABLE, 0, 0 );
+	if( host.type != HOST_DEDICATED )
+		return;
+
+	s_wcd.inputEnabled = false;
 }
 
-void Wcon_SetInputText( const char *inputText )
+static void Wcon_SetInputText( const char *inputText )
 {
-	if( host.type != HOST_DEDICATED ) return;
-	SetWindowText( s_wcd.hwndInputLine, inputText );
-	SendMessage( s_wcd.hwndInputLine, EM_SETSEL, Q_strlen( inputText ), -1 );
+	if( host.type != HOST_DEDICATED )
+		return;
+
+	while( s_wcd.consoleTextLen-- )
+	{
+		Wcon_PrintInternal( "\b \b", 0 );
```

Comment:
Oh, that annoying bell when you press backspace on empty input... xD

Awesome :D